src/test/groovy/org/gradle/android/CrossVersionOutcomeAndRelocationTest.groovy

```diff
@@ -0,0 +1,421 @@
+        assert gradleVersion instanceof GradleVersion
+        assert androidVersion instanceof VersionNumber
+
+        println "> Using Android plugin $androidVersion"
+        println "> Running with $gradleVersion"
+
+        def originalDir = temporaryFolder.newFolder()
+        new SimpleAndroidApp(originalDir, cacheDir, androidVersion, true).writeProject()
+
+        def relocatedDir = temporaryFolder.newFolder()
+        new SimpleAndroidApp(relocatedDir, cacheDir, androidVersion, true).writeProject()
+
+        def expectedResults = expectedResults(androidVersion, gradleVersion)
+
+        println expectedResults.describe()
+
+        cacheDir.deleteDir()
+        cacheDir.mkdirs()
+
+        withGradleVersion(gradleVersion.version)
+            .withProjectDir(originalDir)
+            .withArguments("assemble", "--build-cache", "--stacktrace")
+            .build()
+
+        when:
+        def result = withGradleVersion(gradleVersion.version)
+            .withProjectDir(relocatedDir)
+            .withArguments("assemble", "--build-cache", "--stacktrace")
+            .build()
+
+        then:
+        expectedResults.verify(result)
+
+        cleanup:
+        originalDir.deleteDir()
+        relocatedDir.deleteDir()
+
+        where:
+        //noinspection GroovyAssignabilityCheck
+        [androidVersion, gradleVersion] << Versions.SUPPORTED_VERSIONS_MATRIX.entries().collect { [it.key, it.value] }
+    }
+
+    static class ExpectedResults {
+        private final Map<String, ExpectedOutcome> outcomes
+
+        ExpectedResults(Map<String, ExpectedOutcome> outcomes) {
+            this.outcomes = outcomes
+        }
+
+        String describe() {
+            "> Expecting ${outcomes.values().count(FROM_CACHE)} tasks out of ${outcomes.size()} to be cached"
+        }
+
+        boolean verify(BuildResult result) {
+            boolean allMatched = true
+            def remainingTasks = result.tasks.collect { it.path }
+            outcomes.each { taskName, expectedOutcome ->
+                def taskOutcome = result.task(taskName)?.outcome
+                if (taskOutcome == null) {
+                    println "> Task '$taskName' was expected to execute but did not"
+                    allMatched = false
+                } else if (expectedOutcome != ExpectedOutcome.UNKNOWN && taskOutcome.name() != expectedOutcome.name()) {
+                    println "> Task '$taskName' was $taskOutcome but should have been $expectedOutcome"
+                    allMatched = false
+                }
+                remainingTasks.remove(taskName)
+            }
+            if (!remainingTasks.empty) {
+                remainingTasks.each { taskName ->
+                    def taskOutcome = result.task(taskName)?.outcome
+                    println "> Task '$taskName' executed with outcome $taskOutcome but was not expected"
+                    allMatched = false
+                }
+            }
+            return allMatched
```

Comment:
I like how this is better readable now. What I'd add on top is an explicit check for deprecation warnings emitted. Just to make sure we don't end up using some deprecated APIs unintentionally.